krm-functions/lib/condkptsdk/design.md

```diff
@@ -0,0 +1,180 @@
+
+The `PopulateOwnResourcesFn` is called for each `for` KRM resource instance.
+
+The sdk will handle from here on the lifecycle of the child resources based on the `OwnType`. The sdk uses the `ownerReference` implemented through an annotation to identify child resource instanaces belonging to a for instance 
+
+signature of the populateOwnResourcesFn:
+
+```golang
+type PopulateOwnResourcesFn func(*fn.KubeObject) ([]*fn.KubeObject, error)
+```
+
+Any fn/controller that has own resources MUST implement the `PopulateOwnResourcesFn`.
+
+### GenerateResourceFn
+
+The `GenerateResourceFn` provides:
+- the `for` object as a first parameter (f the object does not exist a nil pointer is provided)
+- the `watch` and `own` instance resource are provided as a list
+
+The `GenerateResourceFn` function either updates the status or generates spec/(status) based on the resource information it is presented. E.g. NAD uses IP and VLAN and interface KRM to generate the KRM.
+
+signature of the generateResourceFn:
+
+```golang
+type GenerateResourceFn func(*fn.KubeObject, []*fn.KubeObject) (*fn.KubeObject, error)
+```
+
+Any fn/controller MUST implement the `GenerateResourceFn`.
+
+### sdk phases
+
+The SDK operates in phases when being executed within a fn/controller
+- firstly the sdk builds up an inventory based on the filters
+- Secondly the global watch callbacks are called. The fn/controller implenmting these callback use the data for 2 things:
+    1. uses attributes of the KRM for further processing later on
+    2. provide feedback on readiness through the error
+- Afterwards if the readiness gate is true the `PopulateOwnResourcesFn` is called if defined by the fn/controller.
+    - the resources returned are used as 
+        - child resource within the `for` instance
+        - the sdk performs a diff and brings the actual state in line with the desired state
+- In the final phase when readiness is determined, the sdk executes the `GenerateResourceFn`. The result is the final step and the data is added/updated in the resourceList of the kpt package
+
+Each function/controller has to implement `GenerateResourceFn`. Only the functions/controller having own resource have to implement `PopulateOwnResourcesFn`.
+
+### pipeline stages
+
+Right now the kpt pipeline is used to execute the conditional dance
+
+## see also
+
+- [golang](https://go.dev)
+- [configuration as data](https://github.com/GoogleContainerTools/kpt/blob/main/docs/design-docs/06-config-as-data.md)
+- [cad concepts](https://kpt.dev/book/02-concepts/)
+- [kpt](https://kpt.dev/book/02-concepts/03-functions)
+- [porch](https://kpt.dev/guides/porch-user-guide)
+- [kpt fn sdk](https://github.com/GoogleContainerTools/kpt-functions-sdk)
```

Comment:
Thanks for the write up. I think it will be useful to explain this with an example ( not code but in text).  Just one or two paragraphs. Let's say
for interface resource,
1. For 
2. Own
3. Watch
resources, and 
1. how the Interface function registers with the SDK
2. What Call back functions need to be implemented and how they are called
3. How eventually the interface function generated the IP/VLAN/ reources and NAD Condition working with the SDK

Reply:
added examples